akd/src/directory.rs

```diff
@@ -233,68 +244,155 @@ impl<S: Storage + Sync + Send, V: VRFKeyStorage> Directory<S, V> {
 
         let current_azks = self.retrieve_current_azks().await?;
         let current_epoch = current_azks.get_latest_epoch();
+        let lookup_info = self
+            .get_lookup_info::<H>(uname.clone(), current_epoch)
+            .await?;
+        let lookup_proof = self
+            .lookup_with_info::<H>(uname, &current_azks, current_epoch, lookup_info)
+            .await;
+        lookup_proof
+    }
 
+    async fn lookup_with_info<H: Hasher>(
+        &self,
+        uname: AkdLabel,
+        current_azks: &Azks,
+        current_epoch: u64,
+        lookup_info: LookupInfo,
+    ) -> Result<LookupProof<H>, AkdError> {
+        let current_version = lookup_info.value_state.version;
+
+        let lookup_proof = LookupProof {
+            epoch: current_epoch,
+            plaintext_value: lookup_info.value_state.plaintext_val,
+            version: lookup_info.value_state.version,
+            exisitence_vrf_proof: self
+                .vrf
+                .get_label_proof::<H>(&uname, false, current_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            existence_proof: current_azks
+                .get_membership_proof(&self.storage, lookup_info.existent_label, current_epoch)
+                .await?,
+            marker_vrf_proof: self
+                .vrf
+                .get_label_proof::<H>(&uname, false, lookup_info.marker_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            marker_proof: current_azks
+                .get_membership_proof(&self.storage, lookup_info.marker_label, current_epoch)
+                .await?,
+            freshness_vrf_proof: self
+                .vrf
+                .get_label_proof::<H>(&uname, true, current_version)
+                .await?
+                .to_bytes()
+                .to_vec(),
+            freshness_proof: current_azks
+                .get_non_membership_proof(
+                    &self.storage,
+                    lookup_info.non_existent_label,
+                    current_epoch,
+                )
+                .await?,
+        };
+
+        // self.storage.log_metrics(log::Level::Info).await;
+
+        Ok(lookup_proof)
+    }
+
+    // TODO(eoz): Call proof generations async
+    /// Allows efficient batch lookups by preloading necessary nodes for the lookups.
+    pub async fn batch_lookup<H: Hasher>(
+        &self,
+        unames: &[AkdLabel],
+    ) -> Result<Vec<LookupProof<H>>, AkdError> {
+        let current_azks = self.retrieve_current_azks().await?;
+        let current_epoch = current_azks.get_latest_epoch();
+
+        // Take a union of the labels we will need proofs of for each lookup.
+        let mut lookup_labels = Vec::new();
+        let mut lookup_infos = Vec::new();
+        for uname in unames {
+            // Save lookup info for later use.
+            let lookup_info = self
```

Comment:
OK so we're preloading nodes, but we're still iteratively loading this. Can we batch this as well? We should know all the necessary information to retrieve the batch of user value state nodes.

Maybe a follow-up issue perhaps for further optimizations?

Reply:
Nice catch, we can do that! Tracked in #167.